### ISSUE TITLE
AMDGPU should implement TargetLowering::hasAndNot

### ISSUE THREAD

**Issue by arsenm** [llvm/llvm-project#112550]
AMDGPU should implement this optimization hook: https://github.com/llvm/llvm-project/blob/b333edd0d6da744c099ad3ff3b5fbd2d4e4dd45a/llvm/include/llvm/CodeGen/TargetLowering.h#L797

This is available on the SALU for i32 and i64 (s_andn2_b32 and s_andn2_b64).  The implementation therefore should be approximately !N->isDivergent() && (N->getValueType() == i32 || i64).

**Comment by arsenm** [comment]
Similarly, hasBitTest should be implemented (using S_BITCMP0_B32/S_BITCMP1_B32/S_BITCMP0_B64/S_BITCMP1_B64)

**Comment by harrisonGPU** [comment]
Hi @arsenm , could you please give me this opportunity?

**Comment by Shoreshen** [comment]
Hi guys, I found that adding hook of `hasAndNot` and `hasAndNotCompare` will trigger the following generic optimizations (LLVM side) of DAG (CombineDAG):

`hasAndNotCompare`:
    (cond:eq/ne (and X, Y), Y) --> (cond:eq/ne (and (not X), Y), 0)
`hasAndNot`:
    (xor (and (xor X, Y), M), Y) --> (or (and X, M), (and Y, (not M)))
    (select (cond:gt X, -1), A, 0) --> (and (not (sra X, size(X)-1)), A)
    (select (cond:gt X, 0), X, 0) --> (and (not (sra X, size(X)-1)), A)

Just wanna ask that if AMDGPU backend  would benefit from these transfermation??

**Comment by harrisonGPU** [comment]
> Hi guys, I found that adding hook of `hasAndNot` and `hasAndNotCompare` will trigger the following generic optimizations (LLVM side) of DAG (CombineDAG):
> 
> `hasAndNotCompare`: (cond:eq/ne (and X, Y), Y) --> (cond:eq/ne (and (not X), Y), 0) `hasAndNot`: (xor (and (xor X, Y), M), Y) --> (or (and X, M), (and Y, (not M))) (select (cond:gt X, -1), A, 0) --> (and (not (sra X, size(X)-1)), A) (select (cond:gt X, 0), X, 0) --> (and (not (sra X, size(X)-1)), A)
> 
> Just wanna ask that if AMDGPU backend would benefit from these transfermation?? As the number of instruction does not reduce after these transfermation~

When I implemented hasAndNot, I noticed that some instructions increased, while others did not. However, the situation will change, so I think hasAndNot does not benefit these transformations.

**Comment by jayfoad** [comment]
> `hasAndNotCompare`:
> (cond:eq/ne (and X, Y), Y) --> (cond:eq/ne (and (not X), Y), 0)

This will be a slight improvement for SALU where AND instructions set SCC depending on whether the result is 0. On VALU it has no benefit.

> `hasAndNot`:
> (xor (and (xor X, Y), M), Y) --> (or (and X, M), (and Y, (not M)))

This looks like it could help with matching V_BFI (but perhaps we have implemented some clever instruction selection patterns to do this already).

> (select (cond:gt X, -1), A, 0) --> (and (not (sra X, size(X)-1)), A)
> (select (cond:gt X, 0), X, 0) --> (and (not (sra X, size(X)-1)), A)

These probably don't help because we already have decent `select` instructions.

**Comment by harrisonGPU** [comment]
> This looks like it could help with matching V_BFI (but perhaps we have implemented some clever instruction selection patterns to do this already)

Yes, I remember we have implemented this hasAndNot in the tablegen file.

**Comment by Shoreshen** [comment]
> > This looks like it could help with matching V_BFI (but perhaps we have implemented some clever instruction selection patterns to do this already)
> 
> Yes, I remember we have implemented this hasAndNot in the tablegen file.

There for I may think that we can do like this:

```cpp
// override hasAndNotCompare
bool SITargetLowering::hasAndNotCompare(SDValue Op) const {
  // Return false if the operation is divergent, as AND-NOT is a scalar-only
  // instruction.
  if (Op->isDivergent())
    return false;

  EVT VT = Op.getValueType();
  return VT == MVT::i32 || VT == MVT::i64;
}

// override hasAndNot
bool SITargetLowering::hasAndNotCompare(SDValue Op) const {
  return false;
}
```

So that we only enable the `(cond:eq/ne (and X, Y), Y) --> (cond:eq/ne (and (not X), Y), 0)` transformation. Then add some cases like the following:

```ll
define i1 @and_cmp1(i32 inreg %x, i32 inreg %y) {
  %and = and i32 %x, %y
  %cmp = icmp eq i32 %and, %y
  ret i1 %cmp
}
```

**Comment by jayfoad** [comment]
> > This looks like it could help with matching V_BFI (but perhaps we have implemented some clever instruction selection patterns to do this already)
> 
> Yes, I remember we have implemented this hasAndNot in the tablegen file.

But it would be nicer to rely on generic DAG combines and remove our clever selection patterns, if possible.

> ```cpp
> bool SITargetLowering::hasAndNot(SDValue Op) const {
>   return false;
> }
> ```

Well... returning `false` from `hasAndNot` makes me uncomfortable because we _do_ have (SALU) and-not instructions. It would not hurt to implement it correctly.

**Comment by harrisonGPU** [comment]
> Therefore, I may think that we can do like this:

I have already implement it :https://github.com/llvm/llvm-project/pull/112647, I need to think about it.

**Comment by harrisonGPU** [comment]
> Well... returning `false` from `hasAndNot` makes me uncomfortable because we _do_ have (SALU) and-not instructions. It would not hurt to implement it correctly.

Got it, Jay. I’ll explore it further. :-)

**Comment by Shoreshen** [comment]
> > > This looks like it could help with matching V_BFI (but perhaps we have implemented some clever instruction selection patterns to do this already)
> > 
> > Yes, I remember we have implemented this hasAndNot in the tablegen file.
> 
> But it would be nicer to rely on generic DAG combines and remove our clever selection patterns, if possible.
> 
> > ```cpp
> > bool SITargetLowering::hasAndNot(SDValue Op) const {
> >   return false;
> > }
> > ```
> 
> Well... returning `false` from `hasAndNot` makes me uncomfortable because we _do_ have (SALU) and-not instructions. It would not hurt to implement it correctly.
> 
> 

yep, that's true that returning false may mislead...... then maybe we should just handle the hasAndNotcompare function. since the hasNotAnd calls it directly....

**Comment by Shoreshen** [comment]
> > Well... returning `false` from `hasAndNot` makes me uncomfortable because we _do_ have (SALU) and-not instructions. It would not hurt to implement it correctly.
> 
> Got it, Jay. I’ll explore it further. :-)

Hi would you mind to add some more cases relating to the transformation?? some of the examples are as follow, thanks~~

```ll
define i1 @and_cmp1(i32 inreg %x, i32 inreg %y) {
  %and = and i32 %x, %y
  %cmp = icmp eq i32 %and, %y
  ret i1 %cmp
}

define i32 @and_cmp1(i32 inreg %x, i32 inreg %y, i32 inreg %m) {
  %xor = xor i32 %x, %y
  %and = and i32 %xor, %m
  %xor2 = xor i32 %and, %y
  ret i32 %xor2
}

define i32 @and_cmp1(i32 inreg %x, i32 inreg %y) {
  %cmp = icmp sgt i32 %x, -1
  %select = select i1 %cmp, i32 %y, i32 0
  ret i32 %select
}
```

**Comment by harrisonGPU** [comment]
> > > Well... returning `false` from `hasAndNot` makes me uncomfortable because we _do_ have (SALU) and-not instructions. It would not hurt to implement it correctly.
> > 
> > 
> > Got it, Jay. I’ll explore it further. :-)
> 
> Hi would you mind to add some more cases relating to the transformation?? some of the examples are as follow, thanks~~
> 
> define i1 @and_cmp1(i32 inreg %x, i32 inreg %y) {
>   %and = and i32 %x, %y
>   %cmp = icmp eq i32 %and, %y
>   ret i1 %cmp
> }
> 
> define i32 @and_cmp1(i32 inreg %x, i32 inreg %y, i32 inreg %m) {
>   %xor = xor i32 %x, %y
>   %and = and i32 %xor, %m
>   %xor2 = xor i32 %and, %y
>   ret i32 %xor2
> }
> 
> define i32 @and_cmp1(i32 inreg %x, i32 inreg %y) {
>   %cmp = icmp sgt i32 %x, -1
>   %select = select i1 %cmp, i32 %y, i32 0
>   ret i32 %select
> }

Okay, I will try to add lit test for this transformation, thanks :-)

**Comment by lwshanbd** [comment]
Hi @harrisonGPU , are you still working on this issue?
